### PR TITLE
Abbreviation

### PR DESCRIPTION
**Pull request by pertrai1**
[Open in Gitpod](https://gitpod.io/from-referrer/)&nbsp;[know more](https://www.gitpod.io/docs/pull-requests/)

### Describe your change:

- [x] Add an algorithm?
- [ ] Fix a bug or typo in an existing algorithm?
- [ ] Documentation change?

### Checklist:

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Javascript/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized.
- [x] I know that pull requests will not be merged if they fail the automated tests.
- [x] This PR only changes one algorithm file. To ease review, please open separate PRs for separate algorithms.
- [x] All new JavaScript files are placed inside an existing directory.
- [x] All filenames should use the UpperCamelCase (PascalCase) style. There should be no spaces in filenames.
      **Example:**`UserProfile.js` is allowed but `userprofile.js`,`Userprofile.js`,`user-Profile.js`,`userProfile.js` are not
- [x] All new algorithms have a URL in their comments that points to Wikipedia or another similar explanation.
- [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: Dynamic-Programming/Abbreviation.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * @param {String} a
+ * @param {String} b
+ * @returns {Boolean}
+ * @see https://www.hackerrank.com/challenges/abbr/problem - Related problem on HackerRank.
+ */
+export const abbreviation = (a, b) => {
+  const n = a.length
+  const m = b.length
+
+  let dp = Array.from({length: n + 1}, () => Array(m + 1).fill(false))
```
Comment (appgurueu) [inline Dynamic-Programming/Abbreviation.js:29]: Please give this a more descriptive name such as `canAbbreviate`.

File: Dynamic-Programming/Abbreviation.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
+/**
+ * @description
+ * Given two strings, `a` and `b`, determine if it's possible to make `a` equal
+ * to `b` You can perform the following operations on the string `a`:
+ * 1. Capitalize zero or more of `a`'s lowercase letters.
+ * 2. Delete all the remaining lowercase letters in `a`.
+ *
+ * ### Algorithm
```
Comment (appgurueu) [inline Dynamic-Programming/Abbreviation.js:8]: Please add these explanatory comments inside the implementation, at the appropriate places.

File: Dynamic-Programming/Abbreviation.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  const n = a.length
+  const m = b.length
+
+  let dp = Array.from({length: n + 1}, () => Array(m + 1).fill(false))
+  dp[0][0] = true
+
+  for (let i = 0; i < n; i++) {
+    for (let j = 0; j <= m; j++) {
+      if (dp[i][j]) {
+        if (j < m && a[i].toUpperCase() === b[j]) {
```
Comment (appgurueu) [inline Dynamic-Programming/Abbreviation.js:35]: You should explain the rule you employ here.

File: Dynamic-Programming/Abbreviation.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+          dp[i + 1][j + 1] = true
+        }
+        if (a[i] === a[i].toLowerCase()) {
```
Comment (appgurueu) [inline Dynamic-Programming/Abbreviation.js:38]: Same here. Please add a comment.

File: Dynamic-Programming/tests/Abbreviation.test.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import { abbreviation } from '../Abbreviation.js'
+
+describe('Abbreviation', () => {
+  test('it can abbreviate string a to b', () => {
```
Comment (appgurueu) [inline Dynamic-Programming/tests/Abbreviation.test.js:4]: I'd partition the test cases into positives and negatives (two different `test` blocks), then run the individual tests using a helper function `const expectPositive = (word, abbr) => expect(abbreviation(word, abbr)).toBe(true)` to make it somewhat less verbose (matter of taste though).

File: Dynamic-Programming/Abbreviation.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+ * `false`
+ *
+ * Time Complexity: (O(|a|*|b|)) where `|a|` => length of string `a`
+ *
+ * @param {String} a
+ * @param {String} b
+ * @returns {Boolean}
+ * @see https://www.hackerrank.com/challenges/abbr/problem - Related problem on HackerRank.
+ */
+export const abbreviation = (a, b) => {
```
Comment (appgurueu) [inline Dynamic-Programming/Abbreviation.js:25]: I'd probably call this `isAbbrevation`. Also, please give `a` and `b` more descriptive names like `word` and `abbr`.

**Review by appgurueu — CHANGES_REQUESTED** [review]
Mostly there, just some minor things to address.

**Comment by pertrai1** [comment]
@appgurueu - can you please take a look at the changes made based on the code review? Thanks

**Review by appgurueu — APPROVED** [review]

**Review by raklaptudirm — APPROVED** [review]

**Pull request merged**